### DISCUSSION TITLE
How to implement "Add Inflection Points", "Tunnify" and "Harmonize" with menu entries?

### DISCUSSION THREAD

**Discussion by linusromer** [fontforge/fontforge#4879] [Q&A] [answered]
I am playing with the thought of implementing "Add Inflection Points", "Tunnify (Balance)" and "Harmonize" directly in FontForge with menu entries and then submitting an pull request. Before doing so, I have some questions:

1. Does the FontForge team consider these features as useful at all?

My plan is to begin with "Add Inflection Points" because it works in a comparable manner as "Add Extrema". The function call hierarchy for the font view would be: FVMenuAddInflections -> FVAddInflections -> SplineCharAddInflections -> SplineSetAddInflections -> SplineAddInflections

2. I think "Add Inflection Points" should be placed after "Add Extrema". Is there a special system for the MID (menu ID's?): I would have defined `#define MID_AddInflections	2256` because it is the smallest number beginning with 22 that is not already defined.

Besides the font view menu there is also a char view menu and metrics view menu. It's clear that a char view menu entry does make sense for "Add Inflection".

3. But how about a menu entry in the metrics view for "Add Inflection Points"? 

4. I guess the pixmap has to be drawn in "tango" style for 16px times 16 px. Are there any further restrictions?

I will add a translation for "Add Inflection Points" in the *.po files:
en: "Add Points of Inflections"
de: "Wendepunkte hinzufügen"
....

5. Is there something that I am missing?

**Comment by frank-trampe** [comment]
The inflection points would be regular smooth points added wherever there is a horizontal or vertical inflection point on the curve?

**Comment by linusromer** [comment]
I see that my comparison with "Add Extrema" was misleading: No, I meant to add regular smooth points on ANY inflection points on the curve (they do not have to be horizontal or vertical). It just works similar as "Add Extrema" with regards to input and output. 
And I know that there is already `Spline2DFindPointsOfInflection()` but this is only used for displaying inflection points not for adding them.

**Reply by linusromer** [reply]
I should clarify that I have already written a Python plugin for FontForge at [https://github.com/linusromer/curvatura](https://github.com/linusromer/curvatura) that has these features. But now I want to implement the features in C for direct use in FontForge.

**Reply by ctrlcctrlv** [reply]
Hey, even if no FontForge user finds it useful (highly unlikely), Tunni lines and points have been on the MFEKglif wishlist for years (cf. MFEK/glif#19, MFEK/glif#20), and I can much easier convert C code for use from Rust than Python code (as I've already done for Spiro—[MFEK/spiro-inner.rlib](https://github.com/MFEK/spiro-inner.rlib)). Indeed, I was already planning on converting your merge function for use in MFEKglif, so, if you do this too it's just more for me to do at the time I do that.

**Reply by linusromer** [reply]
@ctrlcctrlv I am not sure if the implementation in C helps more for Rust than the implementation in Python. It's probably easier to take the formulae from the Curvatura documentation for a new implementation in Rust.

Meanwhile, I have tried to implement "Add points of inflections" as described above. It seems to work if I select the nodes before (which I will try to fix):
![inflections](https://user-images.githubusercontent.com/11213578/148429195-8cba94f9-b02d-4084-8583-c623d2a5ee6c.png)
This does not necessarily answer my questions 2 to 5. However, I think I can add to the 5th question:
Yes, I did forget scripting (functions for python and ff scripts) and documentation.

**Reply by ctrlcctrlv** [reply]
No, I'm sure it helps more, because of `c2rust`. Your work is quite complex, and MFEK's `Point` struct is FontForge compatible (prev/next, prev_cp (`b`)/next_cp (`a`) and all). It takes little for me to run the C functions through `c2rust`, barely understanding them, then just map my fields to FontForge's.

**Reply by linusromer** [reply]
I have made a [pull request](https://github.com/fontforge/fontforge/pull/4900) assuming the following answers:
2) `MID_AddInflections 2256` is fine
3) A metrics view entry for "Add Inflection Points" does make as much sense as "Add Extrema"
4) Pixmaps are drawn in tango and old style in the size 16x16.

**Comment by linusromer** [comment, marked as answer]
The answers are given in https://github.com/fontforge/fontforge/pull/4900, which is merged. So I will close this question.
